### PR TITLE
Remove run_after_success

### LINKED ISSUE #10680: Remove run_after_success support from prowjob
We will remove `run_after_success` in a future version of prow. These do not work correctly in a wide variety of edge cases, and we do not want/intent to ever support them.

In the future prow will support scheduling a single prowjob and waiting for it to complete, and this will result in scheduling a single pod/build/pipeline and waiting for it to complete.

People who need to schedule a tree of jobs should create a job of `knative-build` or in the near future `knative-pipeline` https://github.com/knative/build-pipeline or some other strategy.

/assign
/cc @stevekuznetsov @cjwagner

### PR DESCRIPTION
**Pull request by stevekuznetsov**
After long deliberation, we have chosen to remove the run_after_success
functionality of Prow. This sort of inter-job relationship should be
handled by writing a pipeline job using knative pipeline builds or by a
custom downstream process. The maintenance burden of this feature far
outweighed the benefits it provided.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @cjwagner @fejta 
/cc @BenTheElder @krzyzacy @munnerz @kargakis 

Fixes #10680 

screenshot from 2019-01-10 10-28-58

Baby McBabyface?

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
screenshot from 2019-01-10 10-31-41

**Comment by munnerz** [comment]
We make pretty extensive use of this in our Prow deployment currently for our [Tarmak](https://github.com/jetstack/tarmak) project, to reduce the number of VMs we have to provision as presubmits (e2e's can take a while when we test a full build matrix of configurations).

I'm completely in favour of moving to knative build pipelines, or whatever other solution we can, but would only request that we wait until a) recommendations for this exist (afaik, knative build pipelines are still WIP), and that there is some kind of deprecation period too, else I'll be stuck cherry-picking until pipeline support lands & we get around to setting up knative 😬

**Comment by stevekuznetsov** [comment]
@fejta is in the middle of pipeline support with #10666 

I think the lag between this landing and the pipeline support will be short.

**Comment by stevekuznetsov** [comment]
@munnerz if there are major features currently in flight for Prow you're hoping to get but will need to cherry-pick, I will gladly cherry-pick those for you in order to help us land this

**Comment by cjwagner** [comment]
@munnerz Knative Builds work today using the build controller: https://github.com/kubernetes/test-infra/tree/master/prow/cmd/build
That could be a good replacement for `run_after_success` as well.

**Comment by stevekuznetsov** [comment]
@fejta @cjwagner the changes to the security repo generator mean that I'll need to remove the generated `run_after_success` jobs as well in this PR -- is that ok?

**Comment by fejta** [comment]
James:
* You can use knative builds today (and you can deploy knative build without the rest of knative)
* Will have pipeline support soon
* If neither of these work, can we provide you with a utility to stat all jobs at the same time but coordinate when each should really start via gcs?

**Comment by munnerz** [comment]
I'll take a look into the Knative functionality in Prow today and see what we can do with it 😄

I've not been keeping up with developments recently, but we'll look into moving across as soon as possible. It's not like we don't already have a fork [anyway](https://github.com/jetstack/test-infra/commit/c37d8b3620bbd74befa78a1788a3822d3925ed83) 🙃

**Comment by fejta** [comment]
Ideally we can figure out what to do so you don't need to maintain a fork long-term.

**Comment by stevekuznetsov** [comment]
I've de-forked myself and aside from the occasional need to bug someone to push `label_sync` image it's pretty great :)

**Comment by munnerz** [comment]
I was so close to being able to get rid of the fork until that one rules_nodejs issue!

**Comment by BenTheElder** [comment]
we have to remove a few of these ourselves, not too major though.

**Comment by stevekuznetsov** [comment]
Rebased on top of @krzyzacy changes. @BenTheElder  @fejta any updates on what work is left to remove the existing `run_after_success` jobs?

**Comment by BenTheElder** [comment]
yes, we've made progress on the kubeadm presubmit, but it's not finished yet. there's a bit more to untangle.

**Comment by neolit123** [comment]
cc

**Comment by stevekuznetsov** [comment]
@krzyzacy I don't really understand this failure -- it's new and in a function with your name in a TODO so hopefully you can help. My change was to make the test iterate over the list of jobs, not the list of jobs and children, so it should have been (and was until the last rebase) a no-op. Now, I see:

```
--- FAIL: TestPostsubmits (0.00s)
    jobs_test.go:150: Jobs post-sig-storage-local-static-provisioner-build-canary and post-sig-storage-local-static-provisioner-build-release have the same context: 
    jobs_test.go:131: Job post-test-infra-build-smoke needs a context.
    jobs_test.go:150: Jobs post-test-infra-build-smoke and ci-test-infra-bazel have the same context: 
    jobs_test.go:150: Jobs post-test-infra-build-smoke and maintenance-boskos-config-upload have the same context: 
    jobs_test.go:150: Jobs post-test-infra-build-smoke and maintenance-ci-testgrid-config-upload have the same context: 
    jobs_test.go:150: Jobs post-test-infra-build-smoke and post-test-infra-deploy-prow have the same context: 
    jobs_test.go:150: Jobs post-test-infra-build-smoke and post-test-infra-push-prow have the same context: 
    jobs_test.go:150: Jobs ci-test-infra-bazel and maintenance-boskos-config-upload have the same context: 
    jobs_test.go:150: Jobs ci-test-infra-bazel and maintenance-ci-testgrid-config-upload have the same context: 
    jobs_test.go:150: Jobs ci-test-infra-bazel and post-test-infra-deploy-prow have the same context: 
    jobs_test.go:150: Jobs ci-test-infra-bazel and post-test-infra-push-prow have the same context: 
    jobs_test.go:150: Jobs maintenance-boskos-config-upload and maintenance-ci-testgrid-config-upload have the same context: 
    jobs_test.go:150: Jobs maintenance-boskos-config-upload and post-test-infra-deploy-prow have the same context: 
    jobs_test.go:150: Jobs maintenance-boskos-config-upload and post-test-infra-push-prow have the same context: 
    jobs_test.go:150: Jobs maintenance-ci-testgrid-config-upload and post-test-infra-deploy-prow have the same context: 
    jobs_test.go:150: Jobs maintenance-ci-testgrid-config-upload and post-test-infra-push-prow have the same context: 
    jobs_test.go:150: Jobs post-test-infra-deploy-prow and post-test-infra-push-prow have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-build-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-build-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-build-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-build-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-test-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-build-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-build-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-build-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-test-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-10 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-build-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-build-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-test-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-11 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-12 and ci-kubernetes-bazel-build-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-12 and ci-kubernetes-bazel-test-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-12 and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-12 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-12 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-13 and ci-kubernetes-bazel-test-1-10 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-13 and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-13 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-build-1-13 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-10 and ci-kubernetes-bazel-test-1-11 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-10 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-10 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-11 and ci-kubernetes-bazel-test-1-12 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-11 and ci-kubernetes-bazel-test-1-13 have the same context: 
    jobs_test.go:150: Jobs ci-kubernetes-bazel-test-1-12 and ci-kubernetes-bazel-test-1-13 have the same context: 
FAIL
```

Why is this even a failure?

**Comment by stevekuznetsov** [comment]
You added this test in b4b38a110b2ff38b66bc8cb9751c563c27ee9431 and I don't understand why two jobs with different names cannot have the same context to report as. That's exactly how `Presubmit` jobs work right now.

File: config/jobs/kubernetes/test-infra/test-infra-postsubmits.yaml [hunk 1]
```diff
@@ -3,6 +3,7 @@ postsubmits:
   - name: post-test-infra-build-smoke
     agent: knative-build
     decorate: true
+    context: "TODO(SEN): WHAT??"
```
[the remark this answers is not in the document]
Comment (krzyzacy) [reply config/jobs/kubernetes/test-infra/test-infra-postsubmits.yaml]: https://github.com/kubernetes/test-infra/pull/10959/files#diff-5d5b63a73a67b9ae5f871419a2423b43R1159

File: prow/config/config.go [hunk 2]
```diff
@@ -1147,24 +1126,18 @@ func (c *ProwConfig) defaultPresubmitFields(js []Presubmit) {
...
 	}
 }
 
 func (c *ProwConfig) defaultPostsubmitFields(js []Postsubmit) {
 	for i := range js {
 		c.defaultJobBase(&js[i].JobBase)
-		c.defaultPostsubmitFields(js[i].RunAfterSuccess)
-		if js[i].Context == "" {
-			js[i].Context = js[i].Name
-		}
```
Comment (krzyzacy) [inline prow/config/config.go]: I think you want to keep these

**Comment by fejta** [comment]
please pass tests

**Comment by stevekuznetsov** [comment]
@fejta only tests that fail are those that cannot pass until `run_after_success` jobs are actually removed

**Comment by fejta** [comment]
sure, no problem. Just will review when tests are passing

**Comment by krzyzacy** [comment]
cc @jlewi - kubeflow also have a run_after_success job

**Comment by stevekuznetsov** [comment]
@krzyzacy is that in `prow.k8s.io` deployment or a separate one?

**Comment by krzyzacy** [comment]
> is that in prow.k8s.io deployment or a separate one?

yes..
https://github.com/kubernetes/test-infra/blob/58b41bfa9d9946c3d44b19ae3419290725ef1222/config/jobs/kubeflow/kubeflow-presubmits.yaml#L216

**Comment by stevekuznetsov** [comment]
r-r-r-r-r-reeeeebaaaase

**Comment by krzyzacy** [comment]
have fun rebasing :-) also generally speaking we tend to give people a deprecation window for things like this?

Also is there an issue to track the existing run_after_success migration?

**Comment by stevekuznetsov** [comment]
> also generally speaking we tend to give people a deprecation window for things like this?

The reason this is going away is it's impossible to support so we decided it was not necessary.

> Also is there an issue to track the existing run_after_success migration?

I was hoping @BenTheElder @neolit123 could help with this, I don't really understand what is to do

**Comment by neolit123** [comment]
my proposal is the following...

just promote the `pull-kubernetes-e2e-kubeadm-gce` job outside of `run_after_success` and disable it temporary. in the next few days we are going to restart it once we figure out what do we have to do.

i don't know what has to be done. @BenTheElder is saying the we can `build_push`? i'm not sure what this is about but kubernetes-anywhere needs the PR build artifacts from a GCS bucket. we can figure this out and rely on the periodics in the meantime for signal.

sounds good @stevekuznetsov @krzyzacy ?

**Comment by stevekuznetsov** [comment]
As those decisions concern the _content_ of jobs and I do not usually have insight into the ramifications of changing job content for k8s, I will have to defer on whether that sounds good.

**Comment by neolit123** [comment]
> my idea is to move https://github.com/kubernetes/test-infra/blob/master/scenarios/kubernetes_e2e.py#L128-L160 into kubeadm deployer (in up maybe?) so we can query the artifacts after the build step - We can experiment with that.

sure, we can look into solutions.

also i would like to remind that these pre-submit kubeadm jobs have:
```
skip_report: true
```

the actual signal we look at is this one and it's periodic:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-all#kubeadm-gce-master

let's unblock this PR.

**Comment by krzyzacy** [comment]
so it leaves to... @jlewi or @IronPan - can we move the kubeflow pipeline job out of run_after_success? We can potentially migrate it to knative build agent now if you still need build steps? @fejta probably can also help you migrate :-)

**Comment by fejta** [comment]
anything using run after success can now just use a knative-build job to accomplish the same thing.


```yaml
  - name: kubeflow-pipeline-build-image
    always_run: true
    decorate: true
    labels:
      preset-service-account: "true"
    spec:
      containers:
      - image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
        command:
        - ./test/presubmit-tests.sh
        args:
        - --workflow_file
        - build_image.yaml
    run_after_success:
    - name: kubeflow-pipeline-e2e-test
      always_run: true
      decorate: true
      labels:
        preset-service-account: "true"
      spec:
        containers:
        - image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
          command:
          - ./test/presubmit-tests-with-pipeline-deployment.sh
          args:
          - --workflow_file
          - e2e_test_gke_v2.yaml
          - --test_result_folder
          - e2e_test
    - name: kubeflow-pipeline-sample-test
      always_run: true
      decorate: true
      skip_report: true
      labels:
        preset-service-account: "true"
      spec:
        containers:
        - image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
          command:
          - ./test/presubmit-tests-with-pipeline-deployment.sh
          args:
          - --workflow_file
          - sample_test_v2.yaml
          - --test_result_folder
          - sample_test
          - --timeout
          - "3600"
    - name: kubeflow-pipeline-e2e-test-gce-minikube
      always_run: true
      decorate: true
      skip_report: true
      labels:
        preset-service-account: "true"
      spec:
        containers:
        - image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
          command:
          - ./test/presubmit-tests-gce-minikube.sh
          args:
          - --workflow_file
          - e2e_test_gke.yaml
          - --test_result_folder
          - e2e_test_gke
```

becomes
```yaml
  - name: kubeflow-pipeline-presubmit
    always_run: true
    decorate: true
    agent: knative-build
    build_spec:
      steps:
      - name: kubeflow-pipeline-build-image
        image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
        command:
        - ./test/presubmit-tests.sh
        args:
        - --workflow_file
        - build_image.yaml
      - name: kubeflow-pipeline-e2e-test
image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
        command:
        - ./test/presubmit-tests-with-pipeline-deployment.sh
        args:
        - --workflow_file
        - e2e_test_gke_v2.yaml
        - --test_result_folder
        - e2e_test
    - name: kubeflow-pipeline-sample-test
       image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
       command:
       - ./test/presubmit-tests-with-pipeline-deployment.sh
       args:
       - --workflow_file
       - sample_test_v2.yaml
       - --test_result_folder
       - sample_test
       - --timeout
       - "3600"
    - name: kubeflow-pipeline-e2e-test-gce-minikube
       image: gcr.io/k8s-testimages/kubekins-e2e:v20190125-3d9554697e-1.10
       command:
       - ./test/presubmit-tests-gce-minikube.sh
       args:
       - --workflow_file
       - e2e_test_gke.yaml
       - --test_result_folder
       - e2e_test_gke
```

@BenTheElder we can use the same pattern for the kubeadm jobs

**Comment by krzyzacy** [comment]
So looks like 4 jobs will become one job for build agent - do we handle things like merging artifacts now?

**Comment by fejta** [comment]
I don't think I understand question, what are merging artifacts?

**Comment by krzyzacy** [comment]
aka if each build step dumps out a $ARTIFACTS/junit.xml, what's the current behavior?

**Comment by stevekuznetsov** [comment]
@fejta this passes tests now :)

**Comment by krzyzacy** [comment]
> right now Gubernator merges jUnit from all the files it found, so if each step makes a file it should be fine?

@stevekuznetsov I think gcs will overwrite previous junit file if the file names are identical.. We'll see if people complains :-)

**Comment by stevekuznetsov** [comment]
:shipit:

**Comment by cjwagner** [comment]
Wooooo
/meowvie
/joke

**Comment by krzyzacy** [comment]
sent #11045 and #11046

**Pull request merged**